### PR TITLE
Added clustering support for TLS templates

### LINKED ISSUE #2440: Add support for TLS clustering

### PR DESCRIPTION
**Pull request by Ice3man543**
## Proposed changes

Closes #2440

```bash
./nuclei -t ./report -u docs.hackerone.com:443 

                     __     _
   ____  __  _______/ /__  (_)
  / __ \/ / / / ___/ / _ \/ /
 / / / / /_/ / /__/ /  __/ /
/_/ /_/\__,_/\___/_/\___/_/   v2.8.7

                projectdiscovery.io

[INF] Using Nuclei Engine 2.8.7 (latest)
[INF] Using Nuclei Templates 9.3.4 (latest)
[INF] Templates added in last update: 15
[INF] Templates loaded for scan: 2
[INF] Targets loaded for scan: 1
[INF] Templates clustered: 2 (Reduced 1 Requests)
[ssl-issuer] [ssl] [info] docs.hackerone.com [Let's Encrypt]
```

Default run leads to reduction of 6 requests per public templates.

```bash
 ./nuclei -t ~/nuclei-templates/ssl -u docs.hackerone.com:443 

                     __     _
   ____  __  _______/ /__  (_)
  / __ \/ / / / ___/ / _ \/ /
 / / / / /_/ / /__/ /  __/ /
/_/ /_/\__,_/\___/_/\___/_/   v2.8.7

                projectdiscovery.io

[INF] Using Nuclei Engine 2.8.7 (latest)
[INF] Using Nuclei Templates 9.3.4 (latest)
[INF] Templates added in last update: 15
[INF] Templates loaded for scan: 8
[INF] Targets loaded for scan: 1
[INF] Templates clustered: 7 (Reduced 6 Requests)
[ssl-dns-names] [ssl] [low] docs.hackerone.com [docs.hackerone.com]
[tls-version] [ssl] [low] docs.hackerone.com [tls13]
[ssl-issuer] [ssl] [low] docs.hackerone.com [Let's Encrypt]
```

## Checklist

- [ ] Pull request is created against the [dev](https://github.com/projectdiscovery/nuclei/tree/dev) branch
- [ ] All checks passed (lint, unit/integration/regression tests etc.) with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: v2/pkg/templates/cluster.go [hunk 1]
```diff
@@ -78,6 +79,17 @@ func Cluster(list map[string]*Template) [][]*Template {
...
+			for otherKey, other := range list {
+				if len(other.RequestsSSL) == 0 || len(other.RequestsSSL) > 1 {
+					continue
+				}
+				if template.RequestsSSL[0].CanCluster(other.RequestsSSL[0]) {
+					delete(list, otherKey)
+					cluster = append(cluster, other)
+				}
+			}
+		}
```
Comment (tarunKoyalwar) [inline v2/pkg/templates/cluster.go:92]: @Ice3man543 , Thinking out load, This does have any impact . But I was thinking maybe we can check `templateType` when iterating over `list` . and if possible maybe we can use `switch` statements over nested if else

**Review by tarunKoyalwar — APPROVED** [review]
```console
$  ./nuclei -t ~/nuclei-templates/ssl -u scanme.sh

                     __     _
   ____  __  _______/ /__  (_)
  / __ \/ / / / ___/ / _ \/ /
 / / / / /_/ / /__/ /  __/ /
/_/ /_/\__,_/\___/_/\___/_/   v2.8.7

		projectdiscovery.io

[INF] Using Nuclei Engine 2.8.7 (latest)
[INF] Using Nuclei Templates 9.3.4 (latest)
[INF] Templates added in last update: 15
[INF] Templates loaded for scan: 8
[INF] Targets loaded for scan: 1
[INF] Templates clustered: 7 (Reduced 6 Requests)
[mismatched-ssl] [ssl] [low] scanme.sh
[self-signed-ssl] [ssl] [low] scanme.sh
[ssl-issuer] [ssl] [low] scanme.sh [pd]
[tls-version] [ssl] [low] scanme.sh [tls13]
[deprecated-tls] [ssl] [info] scanme.sh [tls10]
[deprecated-tls] [ssl] [info] scanme.sh [tls11]
```

```console
$ ./nuclei -t ~/nuclei-templates/ssl -u scanme.sh -disable-clustering

                     __     _
   ____  __  _______/ /__  (_)
  / __ \/ / / / ___/ / _ \/ /
 / / / / /_/ / /__/ /  __/ /
/_/ /_/\__,_/\___/_/\___/_/   v2.8.7

		projectdiscovery.io

[INF] Using Nuclei Engine 2.8.7 (latest)
[INF] Using Nuclei Templates 9.3.4 (latest)
[INF] Templates added in last update: 15
[INF] Templates loaded for scan: 8
[INF] Targets loaded for scan: 1
[tls-version] [ssl] [info] scanme.sh [tls13]
[mismatched-ssl] [ssl] [low] scanme.sh
[deprecated-tls] [ssl] [info] scanme.sh [tls10]
[ssl-issuer] [ssl] [info] scanme.sh [pd]
[deprecated-tls] [ssl] [info] scanme.sh [tls11]
[self-signed-ssl] [ssl] [low] scanme.sh

```

**Review by ehsandeep — CHANGES_REQUESTED** [review]
```console
cat h1.txt | ./nuclei -pt ssl 

                     __     _
   ____  __  _______/ /__  (_)
  / __ \/ / / / ___/ / _ \/ /
 / / / / /_/ / /__/ /  __/ /
/_/ /_/\__,_/\___/_/\___/_/   v2.8.7

		projectdiscovery.io

[INF] Using Nuclei Engine 2.8.7 (latest)
[INF] Using Nuclei Templates 9.3.4 (latest)
[INF] Templates added in last update: 15
[INF] Templates loaded for scan: 8
[INF] Targets loaded for scan: 21
[INF] Templates clustered: 7 (Reduced 126 Requests)
[ssl-issuer] [ssl] [medium] docs.hackerone.com [Let's Encrypt]
[ssl-dns-names] [ssl] [medium] docs.hackerone.com [docs.hackerone.com]
[tls-version] [ssl] [medium] docs.hackerone.com [tls13]
[ssl-issuer] [ssl] [medium] mta-sts.hackerone.com [Let's Encrypt]
[ssl-dns-names] [ssl] [medium] mta-sts.hackerone.com [mta-sts.hackerone.com]
[tls-version] [ssl] [medium] mta-sts.hackerone.com [tls13]
[ssl-issuer] [ssl] [medium] mta-sts.forwarding.hackerone.com [Let's Encrypt]
[ssl-dns-names] [ssl] [medium] mta-sts.forwarding.hackerone.com [mta-sts.forwarding.hackerone.com]
[tls-version] [ssl] [medium] mta-sts.forwarding.hackerone.com [tls13]
```

Severity information is being wrongly populated randomly.

File: v2/pkg/templates/cluster.go (hunk 1 above)
Comment (Ice3man543) [reply v2/pkg/templates/cluster.go]: Done, let me know if the changes look good!

Comment (tarunKoyalwar) [reply v2/pkg/templates/cluster.go]: @Ice3man543 , lgtm!

**Review by ehsandeep — APPROVED** [review]
```console
echo 66.96.146.129 | ./nuclei -id tls-version,ssl-issuer

                     __     _
   ____  __  _______/ /__  (_)
  / __ \/ / / / ___/ / _ \/ /
 / / / / /_/ / /__/ /  __/ /
/_/ /_/\__,_/\___/_/\___/_/   v2.8.7

		projectdiscovery.io

[INF] Using Nuclei Engine 2.8.7 (latest)
[INF] Using Nuclei Templates 9.3.4 (latest)
[INF] Templates added in last update: 15
[INF] Templates loaded for scan: 2
[INF] Targets loaded for scan: 1
[INF] Templates clustered: 2 (Reduced 1 Requests)
[tls-version] [ssl] [info] 66.96.146.129 [tls12]
[ssl-issuer] [ssl] [info] 66.96.146.129 [Sectigo Limited]
```

Results into single server hello, instead two (running with `-dc` option)

image

**Pull request merged**